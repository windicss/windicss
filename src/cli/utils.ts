import fs from 'fs';
import path from 'path';
import glob from 'glob';
import minimatch from 'minimatch';

export function globArray(patterns: string[], options?: glob.IOptions): string[] {
  const list: string[] = [];

  patterns.forEach(pattern => {
    if (pattern[0] === '!') {
      let i = list.length-1;
      while(i > -1) {
        if (!minimatch(list[i], pattern)) {
          list.splice(i, 1);
        }
        i--;
      }

    }
    else {
      const newList = glob.sync(pattern, options);
      newList.forEach(item => {
        if (list.indexOf(item)===-1) {
          list.push(item);
        }
      });
    }
  });

  return list;
}

export function getVersion(): string {
  // eslint-disable-next-line quotes
  return `__NAME__ __VERSION__`; // replace by rollup
}

export function fuzzy(content: string): string[] {
  return content.match(/([^<>"'`\s]*[^<>"'`\s:])|([^<>"'`\s.(){}[\]#=%]*[^<>"'`\s.(){}[\]#=%:])/g) ?? [];
}

export function generateTemplate(
  folder: string,
  outputPath = 'windi.css'
): { html: string; css: string } {
  if (!(fs.existsSync(folder) && fs.lstatSync(folder).isDirectory())) {
    fs.mkdirSync(folder);
    if (!fs.existsSync(folder))
      throw new Error(`Folder ${folder} creation failed.`);
  }
  folder = path.resolve(folder);
  const template = `<!DOCTYPE html>
  <html lang="en">

  <head>
      <meta charset="UTF-8">
      <meta name="viewport" content="width=device-width, initial-scale=1.0">
      <title>${path.basename(folder)}</title>
      <link rel="stylesheet" href="${outputPath}">
  </head>

  <body class="bg-gray-100">
      <div class="container mx-auto flex flex-col justify-center items-center h-screen">
          <div class="lg:flex shadow rounded-lg">
              <div class="bg-blue-500 rounded-t-lg lg:rounded-tr-none lg:rounded-l-lg lg:w-4/12 py-4 h-full flex flex-col justify-center">
                  <div class="text-center tracking-wide">
                      <div class="text-white font-bold text-8xl lg:text-4xl">24</div>
                      <div class="text-white font-normal text-4xl pt-4 lg:pt-0 lg:text-2xl">Sept</div>
                  </div>
              </div>
              <div class="w-full px-1 bg-white py-5 lg:px-2 lg:py-2 tracking-wide">
                  <div class="flex flex-row lg:justify-start justify-center">
                      <div class="text-gray-700 font-light text-sm text-center lg:text-left px-2">
                          1:30 PM
                      </div>
                      <div class="text-gray-700 font-light text-sm text-center lg:text-left px-2">
                          Organiser : IHC
                      </div>
                  </div>
                  <div class="text-gray-700 text-xl pb-1 text-center lg:text-left px-2">
                      International Conference Dubai
                  </div>

                  <div class="text-gray-800 font-light text-sm pt-1 text-center lg:text-left px-2">
                      A-142/1, A-142, Ganesh Nagar, Tilak Nagar, New Delhi, 110018
                  </div>
              </div>
              <div class="flex flex-row items-center w-full lg:w-1/3 bg-white lg:justify-end justify-center px-2 py-4 lg:px-0 rounded-lg">
                  <span class="tracking-wider text-gray-600 cursor-pointer bg-gray-100 hover:bg-gray-200 px-4 py-2 text-sm rounded-lg leading-loose mx-2">
                      Going
                  </span>
              </div>
          </div>
      </div>
  </body>
  </html>`;
  const inputPath = path.join(folder, 'index.html');
  outputPath = path.join(folder, outputPath);
  fs.writeFileSync(inputPath, template);
  fs.writeFileSync(outputPath, '');
  return { html: inputPath, css: outputPath };
}
