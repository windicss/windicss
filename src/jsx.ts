/* eslint-disable @typescript-eslint/no-empty-interface */
export type UtilityNames =
  | 'p'
  | 'm'
  | 'w'
  | 'h'
  | 'z'
  | 'border'
  | 'grid'
  | 'flex'
  | 'bg'
  | 'text'
  | 'font'
  | 'opacity'
  | 'animate'
  | 'transition'
  | 'transform'
  | 'align'
  | 'justify'
  | 'items'
  | 'block'
  | 'content'
  | 'pos'
  | 'box'
  | 'overflow'
  | 'underline'
  | 'list'
  | 'gradient'
  | 'divide'
  | 'gap'
  | 'ring'
  | 'icon'
  | 'container'
  | 'space'
  | 'table'
  | 'order'
  | 'place'
  | 'display'
  | 'shadow'
  | 'blend'
  | 'filter'
  | 'backdrop'
  | 'cursor'
  | 'outline'
  | 'select'

export type VariantNames =
  | 'hover'
  | 'active'
  | 'focus'
  | 'enabled'
  | 'dark'
  | 'light'
  | 'sm'
  | 'lg'
  | 'md'
  | 'xl'
  | 'xxl'
  | 'first'
  | 'last'
  | 'child'
  | 'root'
  | 'before'
  | 'after'
  | 'all'

export type AttributifyNames<Prefix extends string = ''> =
  | `${Prefix}${UtilityNames}`
  | `${Prefix}${VariantNames}`
  | `${Prefix}${VariantNames}:${UtilityNames}`

export interface AttributifyAttributes extends Partial<Record<AttributifyNames, string>> {}
