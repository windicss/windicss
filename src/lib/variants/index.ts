import { generateOrientations } from './orientation';
import { generateScreens } from './screen';
import { generateThemes } from './theme';
import { generateStates } from './state';

import type { Style } from '../../utils/style';
import type { BaseTheme, Config, DictStr } from '../../interfaces';

export type Variants = {
  orientation: { [key: string]: () => Style };
  screen: { [key: string]: () => Style };
  theme: { [key: string]: () => Style };
  state: { [key: string]: () => Style };
}

export function resolveVariants(config: Config): Variants {
  return {
    orientation: generateOrientations(((config.theme as BaseTheme)?.orientation ?? {})),
    screen: generateScreens(((config.theme as BaseTheme)?.screens ?? {}) as DictStr),
    theme: generateThemes(config.darkMode),
    state: generateStates(config.variantOrder ?? []),
  };
}

export {
  generateOrientations,
  generateScreens,
  generateThemes,
  generateStates,
};
