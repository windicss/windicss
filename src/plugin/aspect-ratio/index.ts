import plugin from '../index';

export default plugin(
  function ({ addUtilities, addDynamic, theme, variants }) {
    addUtilities({
      '.aspect-none': {
        position: 'static',
        paddingBottom: '0',
        '> *': {
          position: 'static',
          height: 'auto',
          width: 'auto',
          top: 'auto',
          right: 'auto',
          bottom: 'auto',
          left: 'auto',
        },
      },
    }, { layer: 'components' });
    addUtilities({
      '.aspect-auto': {
        'aspect-ratio': 'auto',
      },
    }, { layer: 'components' });
    addUtilities({
      '.aspect-square': {
        'aspect-ratio': '1 / 1',
      },
    }, { layer: 'components' });
    addUtilities({
      '.aspect-video': {
        'aspect-ratio': '16 / 9',
      },
    }, { layer: 'components' });
    addDynamic('aspect-w', ({ Utility, Style }) => {
      const prop = Utility.handler
        .handleStatic(theme('aspectRatio'))
        .handleNumber(1, undefined, 'float')
        .value;
      if (!prop) return;
      return Style.generate(Utility.class, {
        '--tw-aspect-w': prop,
        position: 'relative',
        paddingBottom: 'calc(var(--tw-aspect-h) / var(--tw-aspect-w) * 100%)',
        '> *': {
          position: 'absolute',
          height: '100%',
          width: '100%',
          top: '0',
          right: '0',
          bottom: '0',
          left: '0',
        },
      });
    }, {
      layer: 'components',
      variants: variants('aspectRatio'),
      group: 'aspectRatio',
      completions: [
        'aspect-w-{float}',
      ],
    });

    addDynamic('aspect-h', ({ Utility }) => {
      return Utility.handler
        .handleStatic(theme('aspectRatio'))
        .handleNumber(1, undefined, 'float')
        .createProperty('--tw-aspect-h');
    }, {
      layer: 'components',
      variants: variants('aspectRatio'),
      group: 'aspectRatio',
      completions: [
        'aspect-h-{float}',
      ],
    });

    addDynamic('aspect', ({ Utility, Style }) => {
      // aspect-h/w
      const value = Utility.handler.handleFraction().value;
      if (value) {
        return Style.generate(Utility.class, {
          position: 'relative',
          paddingBottom: value,
          '> *': {
            position: 'absolute',
            height: '100%',
            width: '100%',
            top: '0',
            right: '0',
            bottom: '0',
            left: '0',
          },
        });
      }
    }, {
      layer: 'components',
      group: 'aspectRatio',
      completions: [
        'aspect-{fraction}',
      ],
    });
  },
  {
    theme: {
      aspectRatio: {
      },
    },
    variants: {
      aspectRatio: ['responsive'],
    },
  }
);
