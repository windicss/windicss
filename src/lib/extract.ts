import { Utility } from './utilities/handler';
import { deepCopy } from '../utils/tools';
import { Style, Property } from '../utils/style';
import { pluginOrder } from '../config/order';
import { staticUtilities, dynamicUtilities } from './utilities';
import type { Processor } from './index';

export function generateStaticStyle(processor: Processor, className:string, addComment = false): Style | undefined {
  // eslint-disable-next-line no-prototype-builtins
  if (!staticUtilities.hasOwnProperty(className))
    return;

  const style = new Style('.' + className);
  const comment = addComment ? className : undefined;
  const { utility, meta } = staticUtilities[className];
  for (const [key, value] of Object.entries(utility)) {
    style.add(
      Array.isArray(value)
        ? value.map((i) => new Property(key, i, comment))
        : new Property(key, value, comment)
    );
  }
  if (processor._plugin.core && !processor._plugin.core[meta.group]) return;
  return style.updateMeta('utilities', meta.group, pluginOrder[meta.group as keyof typeof pluginOrder], meta.order, true);
}

export default function extract(
  processor: Processor,
  className: string,
  addComment = false,
  prefix?: string,
): Style | Style[] | undefined {
  if (prefix && !className.startsWith(prefix)) return;

  // handle static base utilities
  if (!prefix && className in staticUtilities) return generateStaticStyle(processor, className, addComment);
  if (prefix && className.startsWith(prefix)) {
    className = className.replace(new RegExp(`^${prefix}`), '');
    if (className in staticUtilities) return generateStaticStyle(processor, className, addComment);
  }
  // handle static plugin utilities & components
  const staticPlugins = { ...processor._plugin.utilities, ...processor._plugin.components, ...processor._plugin.shortcuts };
  if (className in staticPlugins) return deepCopy(staticPlugins[className]);

  const utility = new Utility(className, processor._handler);


  // handle dynamic plugin utilities
  for (const [key, generator] of Object.entries(processor._plugin.dynamic)) {
    if (className.match(new RegExp(`^-?${key}`))) {
      let style = generator(utility);
      if (style instanceof Property) style = style.toStyle(utility.class);
      if (style && addComment)
        Array.isArray(style)
          ? style.map((i) => i.property.forEach((p) => (p.comment = className)))
          : style.property.forEach((p) => (p.comment = className));
      if (style) return style;
    }
  }

  // handle dynamic base utilities
  const matches = className.match(/\w+/);
  const key = matches ? matches[0] : undefined;
  // eslint-disable-next-line no-prototype-builtins
  if (key && dynamicUtilities.hasOwnProperty(key)) {
    let style = dynamicUtilities[key](utility, processor.pluginUtils);
    if (!style) return;
    if (processor._plugin.core && !processor._plugin.core[Array.isArray(style) ? style[0].meta.group : style.meta.group]) return;
    if (style instanceof Property) style = style.toStyle(utility.class);
    if (addComment) Array.isArray(style)? style.map((i) => i.property.forEach((p) => (p.comment = className))): style.property.forEach((p) => (p.comment = className));
    return style;
  }
}
