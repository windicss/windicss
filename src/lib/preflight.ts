import { baseUtilities } from './utilities';
import { Style, Property, StyleSheet } from '../utils/style';
import type { ThemeUtil } from '../interfaces';
import type { Processor } from './index';

export default function preflight(
  processor: Processor,
  html?: string,
  includeBase = true,
  includeGlobal = true,
  includePlugins = true,
): StyleSheet {
  // Generate preflight style based on html tags.
  const globalSheet = new StyleSheet();
  const styleSheet = new StyleSheet();
  const pluginSheet = new StyleSheet();

  const createStyle = (
    selector: string | undefined,
    properties: {
      [key: string]: string | string[] | ((theme: ThemeUtil) => string);
    },
    isGlobal = false
  ) => {
    const style = new Style(selector, undefined, false);
    for (const [key, value] of Object.entries(properties)) {
      style.add(
        Array.isArray(value)
          ? value.map((v) => new Property(key, v))
          : new Property(key, typeof value === 'function' ? value((path: string, defaultValue?: unknown) => processor.theme(path, defaultValue)) : value)
      );
    }
    style.updateMeta('base', 'preflight', 0, isGlobal? 1 : 2, true);
    return style;
  };

  const tags = html ? Array.from(new Set(html.match(/<\w+/g))).map((i) => i.substring(1)) : undefined;

  // handle base style
  includeBase && (processor.config('prefixer') ? baseUtilities : baseUtilities.filter(i => !i.selector || !/::?(webkit-input|-moz|-ms-input)-placeholder$/.test(i.selector))).forEach(p => {
    if (includeGlobal && p.global) {
      // global style, such as * or html, body
      globalSheet.add(createStyle(p.selector, p.properties, true));
    } else if (tags !== undefined) {
      // only generate matched styles
      const includeTags = tags.filter((i) => p.keys.includes(i));
      if (includeTags.length > 0) styleSheet.add(createStyle(p.selector ? p.selector : includeTags.join(', '), p.properties));
    } else {
      // if no tags input, generate all styles
      styleSheet.add(createStyle(p.selector ? p.selector : p.keys.join(', '), p.properties));
    }
  });

  // handle plugin style
  if (includePlugins) {
    // base Styles
    let preflightList: Style[] = [];
    Object.values(processor._plugin.preflights).forEach((styles) => {
      preflightList = preflightList.concat(styles);
    });
    pluginSheet.add(preflightList);

    // always generated styles
    let staticList: Style[] = [];
    Object.values(processor._plugin.static).forEach((styles) => {
      staticList = staticList.concat(styles);
    });
    pluginSheet.add(staticList);
  }

  const result = styleSheet.combine().sort().extend(pluginSheet.combine().sort());
  return includeGlobal ? result.extend(globalSheet.combine().sort(), false) : result;
}
