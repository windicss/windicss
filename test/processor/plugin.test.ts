import { FontSize } from '../../dist/types/interfaces';
import { Property, Style } from '../../src/utils/style';
import { Processor } from '../../src/lib';
import { hex2RGB } from '../../src/utils';
import plugin from '../../src/plugin';

const processor = new Processor();

describe('Plugin Method', () => {
  it('addUtilities', () => {
    const utilities = {
      '.skew-10deg': {
        transform: 'skewY(-10deg)',
      },
      '.skew-15deg': {
        transform: 'skewY(-15deg)',
      },
    };
    expect(processor.addUtilities(utilities).map(i => i.build()).join('\n')).toBe('.skew-10deg {\n  transform: skewY(-10deg);\n}\n.skew-15deg {\n  transform: skewY(-15deg);\n}');
  });

  it('addUtilities with non-class styles', () => {
    const utilities = {
      '.skew-10deg': {
        transform: 'skewY(-10deg)',
      },
      '.skew-15deg': {
        transform: 'skewY(-15deg)',
      },
      '[data-theme]': { '--active': '#0099FF' },
    };
    processor.addUtilities(utilities);
    expect(processor.interpret('skew-10deg').styleSheet.build()).toEqual('.skew-10deg {\n  transform: skewY(-10deg);\n}');
    expect(processor.preflight(undefined, false, false, true).build()).toEqual(
      `[data-theme] {
  --active: #0099FF;
}`);
  });

  it('add duplicated utilities', () => {
    const a = {
      '.btn': {
        padding: '.5rem 1rem',
      },
    };

    const b = {
      '.btn': {
        borderRadius: '.25rem',
        fontWeight: '600',
      },
    };

    const processor = new Processor();
    processor.addUtilities(a);
    processor.addUtilities(b);
    expect(processor.interpret('btn').styleSheet.build()).toMatchSnapshot('css');
  });

  it('add pseudo utilities', () => {
    const utilities = {
      '.btn': {
        padding: '.5rem 1rem',
        borderRadius: '.25rem',
      },
      '.btn:hover': {
        color: '#fff',
      },
      '.hover\\:abc:hover': {
        color: '#1c1c1e',
      },
      '@media (min-width: 640px)': {
        '.hover\\:abc:hover': {
          color: '#fff',
        },
      },
      '@media (min-width: 768px)': {
        '@keyframes ping': {
          '.test': {
            color: 'red',
          },
          '.btn': {
            fontWeight: '600',
          },
        },
      },
    };
    const processor = new Processor();
    processor.addUtilities(utilities);
    expect(processor.interpret('btn hover:abc test').styleSheet.build()).toMatchSnapshot('css');
  });

  it('add multi utilities', () => {
    const utilities = {
      '.a, .b:hover, [type=\'button\']': {
        padding: '.5rem 1rem',
        borderRadius: '.25rem',
      },
    };
    const processor = new Processor();
    processor.addUtilities(utilities);
    expect(processor.preflight(undefined, false, false, true).build()).toMatchSnapshot('preflight');
    expect(processor.interpret('a').styleSheet.build()).toMatchSnapshot('a');
    expect(processor.interpret('b').styleSheet.build()).toMatchSnapshot('b');
  });

  it('add utilities containing array of style objects #439', () => {
    const a = {
      '@media (min-width:1024px)': [
        { '.btn': [{ color: 'red' }, { border: '1rem' }] },
        { '.btn-side': { color: 'white', border: '2rem' } },
      ],
    };
    const processor = new Processor();
    processor.addUtilities(a);
    expect(processor.interpret('btn').styleSheet.build()).toMatchSnapshot('css');
    expect(processor.interpret('btn-side').styleSheet.build()).toMatchSnapshot('css');
  });


  it('addComponents', () => {
    const buttons = {
      '.btn': {
        padding: '.5rem 1rem',
        borderRadius: '.25rem',
        fontWeight: '600',
      },
      '.btn-blue': {
        backgroundColor: '#3490dc',
        color: '#fff',
        '&:hover': {
          backgroundColor: '#2779bd',
        },
      },
      '.btn-red': {
        backgroundColor: '#e3342f',
        color: '#fff',
        '&:hover': {
          backgroundColor: '#cc1f1a',
        },
      },
    };

    expect(processor.addComponents(buttons).map(i => i.build()).join('\n')).toBe('.btn {\n  padding: .5rem 1rem;\n  border-radius: .25rem;\n  font-weight: 600;\n}\n.btn-blue {\n  background-color: #3490dc;\n  color: #fff;\n}\n.btn-blue:hover {\n  background-color: #2779bd;\n}\n.btn-red {\n  background-color: #e3342f;\n  color: #fff;\n}\n.btn-red:hover {\n  background-color: #cc1f1a;\n}');
  });

  it('addComponents #242', () => {
    const processor = new Processor();
    processor.addComponents({
      '.avatar>div': { display: 'block', overflow: 'hidden' },
      '.avatar+div': { display: 'block', overflow: 'hidden' },
      '.avatar div': { display: 'block', overflow: 'hidden' },
      '.avatar~div': { display: 'block', overflow: 'hidden' },
    });
    const css = processor.preflight(undefined, false, false, true).build();
    expect(css).toMatchSnapshot('css');
  });

  it('addComponents with non-class styles ', () => {
    const processor = new Processor();
    const components = {
      '[data-theme]': { '--active': '#0099FF' },
      '[data-theme="dark"]': {
        '--primary': '#111111',
        '--on-primary': '#B4B4B4',
        '--on-primary-active': '#F2F2F2',
        '--frame': '#1E1E1E',
        '--on-frame': '#808080',
        '--on-frame-active': 'var(--on-primary)',
      },
      '[data-theme="light"]': {
        '--primary': '#FFFFFF',
        '--on-primary': '#888888',
        '--on-primary-active': '#333333',
        '--frame': '#EFEFEF',
        '--on-frame': '#808080',
        '--on-frame-active': 'var(--on-primary)',
      },
    };
    processor.addComponents(components);
    const css = processor.preflight('test', false, false, true).build();
    expect(css).toMatchSnapshot('css');
  });

  it('add duplicated components', () => {
    const a = {
      '.btn': {
        padding: '.5rem 1rem',
      },
    };

    const b = {
      '.btn': {
        borderRadius: '.25rem',
        fontWeight: '600',
      },
    };

    const processor = new Processor();
    processor.addComponents(a);
    processor.addComponents(b);
    expect(processor.interpret('btn').styleSheet.build()).toMatchSnapshot('css');
  });

  it('add pseudo components', () => {
    const utilities = {
      '.btn': {
        padding: '.5rem 1rem',
        borderRadius: '.25rem',
      },
      '.btn:hover': {
        color: '#fff',
      },
      '.hover\\:abc:hover': {
        color: '#1c1c1e',
      },
      '@media (min-width: 640px)': {
        '.hover\\:abc:hover': {
          color: '#fff',
        },
      },
      '@media (min-width: 768px)': {
        '@keyframes ping': {
          '.test': {
            color: 'red',
          },
          '.btn': {
            fontWeight: '600',
          },
        },
      },
    };
    const processor = new Processor();
    processor.addComponents(utilities);
    expect(processor.interpret('btn').styleSheet.build()).toMatchSnapshot('css');
  });

  it('add components containing array of style objects #439', () => {
    const a = {
      '@media (min-width:1024px)': [
        { '.btn': [{ color: 'red' }, { border: '1rem' }] },
        { '.btn-side': { color: 'white', border: '2rem' } },
      ],
    };
    const processor = new Processor();
    processor.addComponents(a);
    expect(processor.interpret('btn').styleSheet.build()).toMatchSnapshot('css');
    expect(processor.interpret('btn-side').styleSheet.build()).toMatchSnapshot('css');
  });

  it('interpret order should follow add components order', () => {
    const a = {
      '.btn': {
        padding: '.5rem 1rem',
      },
    };

    const b = {
      '.btn-sm': {
        borderRadius: '.25rem',
        fontWeight: '600',
      },
      '.btn-lg': {
        borderRadius: '.5rem',
        fontWeight: '700',
      },
    };

    const processor = new Processor();
    processor.addComponents(a);
    processor.addComponents(b);
    expect(processor.interpret('btn-lg btn-sm btn').styleSheet.build()).toMatchSnapshot('css');
    expect(processor.interpret('btn-lg btn-sm btn').styleSheet.build()).toEqual(processor.interpret('btn btn-sm btn-lg').styleSheet.build());
  });

  it('addBase', () => {
    const processor = new Processor();
    expect(processor.addBase({
      'h1': { fontSize: (processor.theme('fontSize.2xl') as FontSize)[0] ?? '1.5rem' },
      'h2': { fontSize: (processor.theme('fontSize.xl') as FontSize)[0] ?? '1.25rem' },
      'h3': { fontSize: (processor.theme('fontSize.lg') as FontSize)[0] ?? '1.125rem' },
    }).map(i => i.build()).join('\n')).toBe('h1 {\n  font-size: 1.5rem;\n}\nh2 {\n  font-size: 1.25rem;\n}\nh3 {\n  font-size: 1.125rem;\n}');
  });

  it('add static base styles', () => {
    const processor = new Processor();
    processor.addBase({
      'h1': { fontSize: (processor.theme('fontSize.2xl') as FontSize)[0] ?? '1.5rem' },
      'h2': { fontSize: (processor.theme('fontSize.xl') as FontSize)[0] ?? '1.25rem' },
    });

    expect(processor.preflight(undefined, false, false, true).build()).toEqual('h1 {\n  font-size: 1.5rem;\n}\nh2 {\n  font-size: 1.25rem;\n}');
  });

  it('add base styles containing array of style objects #439', () => {
    const a = {
      'h1': [{ fontSize: '1.5rem' }, { lineHeight: '.2rem' }],
      'h2': { fontSize: '1.5rem', lineHeight: '.2rem' },
    };
    const processor = new Processor();
    const result = processor.addBase(a).map(i => i.build());
    expect(result[0]).toEqual('h1 {\n  font-size: 1.5rem;\n  line-height: .2rem;\n}');
    expect(result[1]).toEqual('h2 {\n  font-size: 1.5rem;\n  line-height: .2rem;\n}');
  });

  it('addVariant pseudoClass', () => {
    const test = new Style('.float-right', new Property('float', 'right'));
    const style = processor.addVariant('hocus', ({ pseudoClass }) => {
      return pseudoClass('hover:focus');
    });
    expect(Array.isArray(style) || style.extend(test).build()).toMatchSnapshot('extend');
    expect(processor.interpret('hocus:float-right').styleSheet.build()).toMatchSnapshot('css');
  });

  it('addVariant modifySelectors', () => {
    const test = new Style('.float-right', new Property('float', 'right'));
    const style = processor.addVariant('disabled', ({ modifySelectors, separator }) => {
      return modifySelectors(({ className }) => {
        return `.${processor.e(`disabled${separator}${className}`)}:disabled`;
      });
    });
    const css = Array.isArray(style) || style.extend(test).build();
    expect(css).toMatchSnapshot('css');
  });

  it('#72 plugin test', () => {
    const processor = new Processor({
      plugins: [
        {
          handler: ({ addUtilities }) => {
            addUtilities({});
          },
          config: {
            theme: {
              extend: {
                colors: {
                  active: 'var(--active)',
                  primary: 'var(--primary)',
                  'on-primary': 'var(--on-primary)',
                  'on-primary-active': 'var(--on-primary-active)',
                  frame: 'var(--frame)',
                  'on-frame': 'var(--on-frame)',
                  'on-frame-active': 'var(--on-frame-active)',
                },
              },
            },
          },
        },
      ],
    });
    expect(processor.theme('colors.white')).toEqual('#fff');
    expect(processor.theme('colors.active')).toEqual('var(--active)');
    expect(processor.interpret('bg-active bg-on-primary-active').styleSheet.build()).toEqual(`.bg-active {
  background-color: var(--active);
}
.bg-on-primary-active {
  background-color: var(--on-primary-active);
}`);
  });

  it('css vars inside css fns stay intact', () => {
    const processor = new Processor({
      plugins: [
        {
          handler: ({ addUtilities }) => {
            addUtilities({});
          },
          config: {
            theme: {
              extend: {
                colors: {
                  primary: 'var(--testvar)',
                  'secondary': 'hsla(145, 40%, 90%, 1)',
                  'secondaryvar': 'hsla(var(--testvar / 1))',
                },
              },
            },
          },
        },
      ],
    });
    expect(processor.theme('colors.primary')).toEqual('var(--testvar)');
    expect(processor.theme('colors.secondary')).toEqual('hsla(145, 40%, 90%, 1)');
    expect(processor.theme('colors.secondaryvar')).toEqual('hsla(var(--testvar / 1))');

    const ssPrimary = processor.interpret('bg-primary').styleSheet.build();
    expect(ssPrimary).toEqual('.bg-primary {\n  background-color: var(--testvar);\n}');

    const ssSecondary = processor.interpret('bg-secondary').styleSheet.build();
    expect(ssSecondary).toEqual('.bg-secondary {\n  --tw-bg-opacity: 1;\n  background-color: rgba(219, 240, 228, var(--tw-bg-opacity));\n}');

    const ssSecondVar = processor.interpret('bg-secondaryvar').styleSheet.build();
    expect(ssSecondVar).toEqual('.bg-secondaryvar {\n  background-color: hsla(var(--testvar / 1));\n}');
  });

  it('syntax for hex colors', () => {
    const processor = new Processor({
      plugins: [
        plugin(function({ addDynamic }) {
          addDynamic('bg', ({ Utility, Style, Property }) => {
            if (/^#([A-Fa-f0-9]{6}|[A-Fa-f0-9]{3})$/.test(Utility.body)) {
              return Style(Utility.class, [
                Property('--tw-bg-opacity', '1'),
                Property(
                  'background-color',
                  `rgba(${hex2RGB(Utility.body)?.join(', ')}, var(--tw-bg-opacity))`
                ),
              ]);
            }
          });
        }),
      ],
    });
    expect(processor.interpret('bg-#1c1c1e').styleSheet.build()).toEqual(
      `.bg-\\#1c1c1e {
  --tw-bg-opacity: 1;
  background-color: rgba(28, 28, 30, var(--tw-bg-opacity));
}`);

  });

  it('plugin extend function', () => {
    const processor = new Processor({
      plugins: [
        plugin(function() {
          return;
        }, {
          theme: {
            extend: {
              // eslint-disable-next-line @typescript-eslint/no-unused-vars
              colors: theme => ({
                'nord0': '#2E3440',
              }),
            },
          },
        }),
      ],
    });
    expect(processor.theme('colors.nord0')).toEqual('#2E3440');
  });

  it('dump config', () => {
    const processor = new Processor({
      plugins: [
        plugin(function() {
          return;
        }, {
          theme: {
            extend: {
              // eslint-disable-next-line @typescript-eslint/no-unused-vars
              colors: theme => ({
                'nord0': '#2E3440',
              }),
            },
          },
        }),
      ],
    });
    const js = processor.dumpConfig();
    const config = eval(js);
    const newProcessor = new Processor(config);
    expect(newProcessor.interpret('bg-nord0').styleSheet.build()).toMatchSnapshot('css');
  });

  it('add nest class name', () => {
    const a = {
      '.btn.loading:before': {
        'borderRadius': '9999px',
        'borderWidth': '2px',
        'height': '1rem',
        'marginRight': '.5rem',
        'width': '1rem',
        'WebkitAnimation': 'spin 2s linear infinite',
        'animation': 'spin 2s linear infinite',
        'content': '""',
        'borderColor': 'transparent currentColor currentColor transparent',
      },
    };

    const processor = new Processor();
    processor.addComponents(a);
    expect(processor.preflight(undefined, false, false, true).build()).toMatchSnapshot('css');
  });
});
