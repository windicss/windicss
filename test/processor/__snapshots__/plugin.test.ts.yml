'Plugin Method / add components containing array of style objects #439 / css / 0': |-
  @media (min-width:1024px) {
    .btn {
      color: red;
      border: 1rem;
    }
  }
'Plugin Method / add components containing array of style objects #439 / css / 1': |-
  @media (min-width:1024px) {
    .btn-side {
      color: white;
      border: 2rem;
    }
  }
Plugin Method / add duplicated components / css / 0: |-
  .btn {
    padding: .5rem 1rem;
    border-radius: .25rem;
    font-weight: 600;
  }
Plugin Method / add duplicated utilities / css / 0: |-
  .btn {
    padding: .5rem 1rem;
    border-radius: .25rem;
    font-weight: 600;
  }
Plugin Method / add multi utilities / a / 1: |-
  .a {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
Plugin Method / add multi utilities / b / 2: |-
  .b:hover {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
Plugin Method / add multi utilities / preflight / 0: |-
  [type='button'] {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
Plugin Method / add nest class name / css / 0: |-
  .btn.loading:before {
    border-radius: 9999px;
    border-width: 2px;
    height: 1rem;
    margin-right: .5rem;
    width: 1rem;
    -webkit-animation: spin 2s linear infinite;
    animation: spin 2s linear infinite;
    content: "";
    border-color: transparent currentColor currentColor transparent;
  }
Plugin Method / add pseudo components / css / 0: |-
  .btn {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
  .btn:hover {
    color: #fff;
  }
  @media (min-width: 768px) {
    @keyframes ping {
      .btn {
        font-weight: 600;
      }
    }
  }
Plugin Method / add pseudo utilities / css / 0: |-
  .btn {
    padding: .5rem 1rem;
    border-radius: .25rem;
  }
  .btn:hover {
    color: #fff;
  }
  .hover\:abc:hover {
    color: #1c1c1e;
  }
  @media (min-width: 640px) {
    .hover\:abc:hover {
      color: #fff;
    }
  }
  @media (min-width: 768px) {
    @keyframes ping {
      .test {
        color: red;
      }
      .btn {
        font-weight: 600;
      }
    }
  }
'Plugin Method / add utilities containing array of style objects #439 / css / 0': |-
  @media (min-width:1024px) {
    .btn {
      color: red;
      border: 1rem;
    }
  }
'Plugin Method / add utilities containing array of style objects #439 / css / 1': |-
  @media (min-width:1024px) {
    .btn-side {
      color: white;
      border: 2rem;
    }
  }
'Plugin Method / addComponents #242 / css / 0': |-
  .avatar>div {
    display: block;
    overflow: hidden;
  }
  .avatar+div {
    display: block;
    overflow: hidden;
  }
  .avatar div {
    display: block;
    overflow: hidden;
  }
  .avatar~div {
    display: block;
    overflow: hidden;
  }
Plugin Method / addComponents with non-class styles  / css / 0: |-
  [data-theme] {
    --active: #0099FF;
  }
  [data-theme="dark"] {
    --primary: #111111;
    --on-primary: #B4B4B4;
    --on-primary-active: #F2F2F2;
    --frame: #1E1E1E;
    --on-frame: #808080;
    --on-frame-active: var(--on-primary);
  }
  [data-theme="light"] {
    --primary: #FFFFFF;
    --on-primary: #888888;
    --on-primary-active: #333333;
    --frame: #EFEFEF;
    --on-frame: #808080;
    --on-frame-active: var(--on-primary);
  }
Plugin Method / addVariant modifySelectors / css / 0: |-
  .disabled\:float-right:disabled {
    float: right;
  }
Plugin Method / addVariant pseudoClass / css / 1: |-
  .hocus\:float-right:hover:focus {
    float: right;
  }
Plugin Method / addVariant pseudoClass / extend / 0: |-
  .float-right:hover:focus {
    float: right;
  }
Plugin Method / dump config / css / 0: |-
  .bg-nord0 {
    --tw-bg-opacity: 1;
    background-color: rgba(46, 52, 64, var(--tw-bg-opacity));
  }
Plugin Method / interpret order should follow add components order / css / 0: |-
  .btn {
    padding: .5rem 1rem;
  }
  .btn-sm {
    border-radius: .25rem;
    font-weight: 600;
  }
  .btn-lg {
    border-radius: .5rem;
    font-weight: 700;
  }
